tools/shared/src/main/scala/org/scalajs/core/tools/linker/analyzer/Analyzer.scala

```diff
@@ -416,47 +416,43 @@ private final class Analyzer(semantics: Semantics,
       if (!allowAddingSyntheticMethods) {
         tryLookupMethod(proxyName)
       } else {
-        /* The lookup for a target method in this code implements the
-         * algorithm defining `java.lang.Class.getMethod`. This mimics how
-         * reflective calls are implemented on the JVM, at link time.
-         *
-         * Caveat: protected methods are not ignored. This can only make an
-         * otherwise invalid reflective call suddenly able to call a protected
-         * method. It never breaks valid reflective calls. This could be fixed
-         * if the IR retained the information that a method is protected.
-         */
-
-        @tailrec
-        def loop(ancestorInfo: ClassInfo): Option[MethodInfo] = {
-          if (ancestorInfo ne null) {
-            ancestorInfo.methodInfos.get(proxyName) match {
-              case Some(m) =>
-                assert(m.isReflProxy && !m.isAbstract)
-                Some(m)
-
-              case _ =>
-                ancestorInfo.findProxyMatch(proxyName) match {
-                  case Some(target) =>
-                    val targetName = target.encodedName
-                    Some(ancestorInfo.createReflProxy(proxyName, targetName))
-
-                  case None =>
-                    loop(ancestorInfo.superClass)
-                }
-            }
-          } else {
-            None
+        methodInfos.get(proxyName).orElse {
+          findReflectiveTarget(proxyName).map { reflectiveTarget =>
+            createReflProxy(proxyName, reflectiveTarget.encodedName)
```

Comment:
Not always. We could, in theory, try and move the declaration as far up as possible, i.e., at the last superclass before an ancestor where it's either not resolvable, or resolves to a different target name. However, this requires re-resolving at every level of the inheritance chain; and in case the declaration can be moved up, we'll need to *redo* the whole process every single time we call that proxy (because we won't have a trace in the current class that we resolved it here already).

Previously, we could get away with a simple scheme where we would define it at the level where we found the target. This was fine with class methods only. But it doesn't work when default methods are introduced.

Therefore, we keep it simple here, if not optimal, and always create the proxies at the leaves, which is safe. Note that this is actually also what we do for default bridges, so far. In the future, we could optimize this, in which case we should probably do it for both default bridges and reflective proxies.